src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -2075,6 +2075,29 @@ private static bool IsEquivalentSymbol(ISymbol declaredSymbol, ISymbol otherSymb
             foreach (IOperation operationBlock in operationBlocks)
             {
                 operationsToAnalyze.AddRange(operationBlock.DescendantsAndSelf());
+
+                // Special handling for IMethodBodyOperation and IConstructorBodyOperation.
+                // These are newly added root operation nodes for C# method and constructor bodies.
+                // However, to avoid a breaking change for existing operation block analyzers,
+                // we have decided to retain the current behavior of making operation block callbacks with the contained
+                // method body and/or constructor initialer operation nodes.
+                // Hence we detect here if the operation block is parented by IMethodBodyOperation or IConstructorBodyOperation ands
+                // add them to 'operationsToAnalyze' so that analyzers that explicitly register for these operation kinds
+                // can get callbacks for these nodes.
+                if (operationBlock.Parent != null)
+                {
+                    switch (operationBlock.Parent.Kind)
+                    {
+                        case OperationKind.MethodBodyOperation:
+                        case OperationKind.ConstructorBodyOperation:
+                            operationsToAnalyze.Add(operationBlock.Parent);
```

Comment:
>operationsToAnalyze.Add(operationBlock.Parent); [](start = 28, length = 47)

Does this mean we are going to add duplicates for constructor that has a body and initializer, or for cases when block body and expression body are present at the same time?

Reply:
No, `operationsToAnalyze` returned by this method are only used for making explicit callbacks, we don't walk down it's descendants. The only place where we walk down descendants of an operation is right at the start of the loop in this method - we invoke `IOperation.DescendantsAndSelf` for every root operation block to determine child operations which need to be called back into the analyzer for `RegisterOperationAction(..., OperationKinds)`

Reply:
> operationsToAnalyze returned by this method are only used for making explicit callbacks

It does look like operationsToAnalyze  is going to have duplicates. Are you saying this doesn't have any negative effect? Could we trigger callback more than once in this case? I would recommend to de-dupe the MethodBodyOperation and ConstructorBodyOperation anyway, it should be easy to do with a local HashSet (probably even pooled).

---
In reply to: [171938354](https://github.com/dotnet/roslyn/pull/25193#discussion_r171938354) [](ancestors = 171938354)

Reply:
> It does look like operationsToAnalyze is going to have duplicates

Ah, you are correct. Let me fix this so we handle it property and also add an assert at the end. Note that using a hashset to keep track of all descendants will cause operation nodes to be called back in different order - while we don't guarantee any ordering for node callbacks, that would be an unnecessary change that might break analyzers depending on this order somehow.